workers/unity/Assets/Playground/Editor/Prefabs/PrefabPreprocessor.cs

```diff
@@ -0,0 +1,122 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Improbable.Gdk.Core.GameObjectRepresentation;
+using UnityEditor;
+using UnityEditor.Build;
+using UnityEditor.Build.Reporting;
+using UnityEngine;
+using Object = UnityEngine.Object;
+
+namespace Playground.Editor
+{
+    [InitializeOnLoad]
+    public class PrefabPreprocessor : IPreprocessBuildWithReport
+    {
+        // Needed for IPreprocessBuildWithReport
+        public int callbackOrder => 0;
+
+        static PrefabPreprocessor()
+        {
+            EditorApplication.playModeStateChanged += PlaymodeStateChanged;
+        }
+
+        void IPreprocessBuildWithReport.OnPreprocessBuild(BuildReport report)
+        {
+            PreprocessPrefabs();
+        }
+
+        private static void PlaymodeStateChanged(PlayModeStateChange playModeStateChange)
+        {
+            if (playModeStateChange == PlayModeStateChange.ExitingEditMode)
+            {
+                PreprocessPrefabs();
+            }
+        }
+
+        private static void PreprocessPrefabs()
+        {
+            if (!DoPrefabsNeedFixing())
+            {
+                return;
+            }
+
+            FixPrefabs();
+
+            AssetDatabase.SaveAssets();
+        }
+
+        private static bool DoPrefabsNeedFixing()
+        {
+            return AssetDatabase.FindAssets("t:Prefab")
+                .Select(AssetDatabase.GUIDToAssetPath)
+                .Select(AssetDatabase.LoadAssetAtPath<GameObject>)
+                .Any(PrefabNeedsFixing);
+        }
+
+        private static bool PrefabNeedsFixing(GameObject prefabObject)
+        {
+            return prefabObject != null
+                && prefabObject.GetComponents<MonoBehaviour>()
+                    .Any(BehaviourNeedsFixing);
+        }
+
+        private static void FixPrefabs()
+        {
+            var prefabsToFix = new List<GameObject>();
+            var monoBehavioursToDisable = new List<MonoBehaviour>();
+
+            var allPrefabObjectsInProject = AssetDatabase.FindAssets("t:Prefab")
+                .Select(AssetDatabase.GUIDToAssetPath)
+                .Select(AssetDatabase.LoadAssetAtPath<GameObject>)
+                .Where(prefabObject => prefabObject != null);
+
+            foreach (var prefabObject in allPrefabObjectsInProject)
+            {
+                var prefabNeedsFixing = false;
+
+                foreach (var monoBehaviour in prefabObject
+                    .GetComponents<MonoBehaviour>()
+                    .Where(BehaviourNeedsFixing))
+                {
+                    if (!prefabNeedsFixing)
+                    {
+                        prefabNeedsFixing = true;
+                        prefabsToFix.Add(prefabObject);
```

Comment:
Why do we use this list only for marking things as dirty? and fix everything anyways

Reply:
we need to mark all these prefabs as dirty only once, but each prefab can contain multiple behaviours to fix

Reply:
But you already have to load all monobehaviours into memory. In the case where a prefab doesn't need fixing you also have to iterate over all monobehaviours on that prefab.

So it will be strictly less work and I think less code to fix all monobehaviours during the first iteration.

Reply:
Simplified greatly